src/TheCompact.sol

```diff
@@ -0,0 +1,240 @@
+    }
+
+    function deposit(address token, address allocator, ResetPeriod resetPeriod, Scope scope, uint256 amount, address recipient) external returns (uint256) {
+        return _performCustomERC20Deposit(token, allocator, resetPeriod, scope, amount, recipient);
+    }
+
+    function deposit(uint256[2][] calldata idsAndAmounts, address recipient) external payable returns (bool) {
+        _processBatchDeposit(idsAndAmounts, recipient);
+
+        return true;
+    }
+
+    function depositAndRegister(uint256[2][] calldata idsAndAmounts, bytes32[2][] calldata claimHashesAndTypehashes, uint256 duration) external payable returns (bool) {
+        _processBatchDeposit(idsAndAmounts, msg.sender);
+
+        return _registerBatch(claimHashesAndTypehashes, duration);
+    }
+
+    function deposit(
+        address token,
+        uint256, // amount
+        uint256, // nonce
+        uint256, // deadline
+        address, // depositor
+        address, // allocator
+        ResetPeriod, // resetPeriod
+        Scope, //scope
+        address recipient,
+        bytes calldata signature
+    ) external returns (uint256) {
+        return _depositViaPermit2(token, recipient, signature);
+    }
+
+    function depositAndRegister(
+        address token,
+        uint256, // amount
+        uint256, // nonce
+        uint256, // deadline
+        address depositor, // also recipient
+        address, // allocator
+        ResetPeriod resetPeriod,
+        Scope, //scope
+        bytes32 claimHash,
+        CompactCategory compactCategory,
+        string calldata witness,
+        bytes calldata signature
+    ) external returns (uint256) {
+        return _depositAndRegisterViaPermit2(token, depositor, resetPeriod, claimHash, compactCategory, witness, signature);
+    }
+
+    function deposit(
+        address, // depositor
+        ISignatureTransfer.TokenPermissions[] calldata permitted,
+        uint256, // nonce
+        uint256, // deadline
+        address, // allocator
+        ResetPeriod, // resetPeriod
+        Scope, //scope
+        address recipient,
+        bytes calldata signature
+    ) external payable returns (uint256[] memory) {
+        return _depositBatchViaPermit2(permitted, recipient, signature);
+    }
+
+    function depositAndRegister(
+        address depositor,
+        ISignatureTransfer.TokenPermissions[] calldata permitted,
+        uint256, // nonce
+        uint256, // deadline
+        address, // allocator
+        ResetPeriod resetPeriod,
+        Scope, //scope
+        bytes32 claimHash,
+        CompactCategory compactCategory,
+        string calldata witness,
+        bytes calldata signature
+    ) external payable returns (uint256[] memory) {
+        return _depositBatchAndRegisterViaPermit2(depositor, permitted, resetPeriod, claimHash, compactCategory, witness, signature);
+    }
+
+    function allocatedTransfer(BasicTransfer calldata transfer) external returns (bool) {
```

Comment:
The functions `allocatedTransfer` as well as `allocatedWithdrawal` always requires the msg.sender to be the sponsor. Why do we not add a `from` address as an input and make use of the granular approval system of ERC 6909? 
So either the msg.sender is the sponsor directly, or the msg.sender has previously received a sufficient token approval from the sponsor. This would provide flexibility and enable a lot of use cases with the ERC6909 tokens, while approval would also be a well recognized and still simple system to handle this.

The signature / approval of the allocator would of course still be required.

Reply:
left a note in "Path To V1" section of the README to explore this idea: https://github.com/Uniswap/the-compact/commit/a5f6b0a553ae7969586b23c463d56e122292520e